test/e2e/oauth2-authorization-code.js

```diff
@@ -0,0 +1,303 @@
+const { exec, fork } = require('child_process');
+const fs = require('fs');
+const path = require('path');
+const url = require('url');
+const util = require('util');
+
+const assert = require('chai').assert;
+const cpr = require('cpr');
+const express = require('express');
+const request = require('superagent');
+const rimraf = require('rimraf');
+const tmp = require('tmp');
+const webdriver = require('selenium-webdriver');
+const chrome = require('selenium-webdriver/chrome');
+const yaml = require('js-yaml');
+
+require('util.promisify/shim')();
+
+const { findOpenPortNumbers, generateBackendServer } =
+  require('../common/server-helper');
+
+const baseConfigDirectory = path.join(__dirname, '..', 'fixtures', 'authorization-code');
+
+describe('oauth2 authorization code grant type', () => {
+  const username = 'kate';
+  const state = 'ABC123xyz';
+
+  let password = null;
+  let clientID = null;
+  let clientSecret = null;
+
+  let testGatewayConfigPath = null;
+  let testGatewayConfigData = null;
+
+  let gatewayProcess = null;
+
+  let gatewayPort = null;
+  let adminPort = null;
+  let backendPort = null;
+  let redirectPort = null;
+
+  let redirectParams = null;
+
+  before(function () {
+    return startGatewayInstance()
+      .then(() => {
+        const args = [
+          '-p', `username=${username}`,
+          '-p', 'firstname=Kate',
+          '-p', 'lastname=Smith'
+        ];
+
+        return createUser(args);
+      })
+      .then(() => {
+        const args = [
+          '-c', username,
+          '-t', 'basic-auth'
+        ];
+
+        return createCredential(args);
+      })
+      .then(credentials => {
+        password = credentials.password;
+
+        const args = [
+          '-u', username,
+          '-p', 'name=apptastic',
+          '-p', `redirectUri=http://localhost:${redirectPort}/cb`
+        ];
+
+        return createApp(args);
+      })
+      .then(app => {
+        const args = [
+          '-c', app.id,
+          '-t', 'oauth2'
+        ];
+
+        return createCredential(args);
+      })
+      .then(credential => {
+        clientID = credential.id;
+        clientSecret = credential.secret;
+      });
+  });
+
+  after(done => {
+    gatewayProcess.kill();
+    rimraf(path.join(testGatewayConfigPath, '..'), done);
+  });
+
+  it('authorizes a valid user', () => {
+    // user login
+    // consent to authorization
+    // verify redirect URI endpoint gets auth code
+    // get access token
+    // make call to backend with access token, verify
+
+    const authURL = url.format({
+      protocol: 'http',
+      hostname: 'localhost',
+      port: gatewayPort,
+      pathname: '/oauth2/authorize',
+      query: {
+        response_type: 'code',
+        client_id: clientID,
+        state: state,
+        redirect_uri: `http://localhost:${redirectPort}/cb`
+      }
+    });
+
+    const driver = new webdriver.Builder()
+      .forBrowser('chrome')
```

Comment:
@kevinswiber chromedriver is a plugin to selenium that operates on real chrome browser. you need to have chrome on it's own installed. PhantomJS is a browser on it's own not a plugin, so when you have phantomjs plugin for selenium that operates against phantomjs browser it will work. And PhantomJS browser can be a dependency